flink-runtime/src/main/scala/org/apache/flink/runtime/jobmanager/JobManager.scala

```diff
@@ -400,20 +400,16 @@ class JobManager(
       import scala.collection.JavaConverters._
       sender ! RegisteredTaskManagers(instanceManager.getAllRegisteredInstances.asScala)
 
-    case Heartbeat(instanceID, metricsReport, accumulators) =>
+    case Heartbeat(instanceID, metricsReport, accumulators, asyncAccumulatorUpdate) =>
       log.debug(s"Received hearbeat message from $instanceID.")
 
-      Future {
-        accumulators foreach {
-          case accumulators =>
-              currentJobs.get(accumulators.getJobID) match {
-                case Some((jobGraph, jobInfo)) =>
-                  jobGraph.updateAccumulators(accumulators)
-                case None =>
-                  // ignore accumulator values for old job
-              }
-        }
-      }(context.dispatcher)
+      if (asyncAccumulatorUpdate) {
```

Comment:
Mixing test code in your production code is IMO bad style. If you need synchronous execution, then start the actor with a `CallingThreadDispatcher` or overwrite this message in the `TestingJobManager`.